invoker shutdown does not take runc into account

## Steps to reproduce the issue:

1. install whisk  
2. create a new action and run it
3.  `docker stop invoker0`


## Provide the expected results and outputs:

```
Shutdown hook should unpause (runc resume) the container and remove it.
```


## Provide the actual results and outputs:

```
[2017-03-08T09:33:36.374Z] [WARN] [??] [ContainerPool] Shutdown hook activated.  Starting container shutdown
[2017-03-08T09:33:36.374Z] [INFO] [#sid_101] [ContainerUtils] [marker:invoker_docker.ps_start:23342545]
[2017-03-08T09:33:36.374Z] [INFO] [#sid_101] [SimpleExec] Running command: /usr/bin/docker ps --no-trunc -a
[2017-03-08T09:33:36.391Z] [INFO] [#sid_101] [SimpleExec] Done running command: /usr/bin/docker ps --no-trunc -a
[2017-03-08T09:33:36.391Z] [INFO] [#sid_101] [ContainerUtils] [marker:invoker_docker.ps_finish:23342564:19]
[2017-03-08T09:33:36.391Z] [INFO] [#sid_101] [ContainerPool] killStragglers now removing 1 leftover containers
[2017-03-08T09:33:36.392Z] [INFO] [#sid_101] [ContainerUtils] [marker:invoker_docker.unpause_start:23342565]
[2017-03-08T09:33:36.392Z] [INFO] [#sid_101] [SimpleExec] Running command: /usr/bin/docker unpause wsk0_177256_limeusibmcomactioninvokenodejs1488965337811001_20170308T092858748Z
[2017-03-08T09:33:36.407Z] [INFO] [#sid_101] [SimpleExec] Done running command: /usr/bin/docker unpause wsk0_177256_limeusibmcomactioninvokenodejs1488965337811001_20170308T092858748Z
[2017-03-08T09:33:36.407Z] [ERROR] [#sid_101] [ContainerUtils] stdout:  stderr: Error response from daemon: Container 480f96278f99592139a19e0cc4ba756f15d74c3052fa2d1524ebdf7edc840e70 is not paused [marker:invoker_docker.unpause_error:23342580:15]
[2017-03-08T09:33:36.407Z] [INFO] [#sid_101] [ContainerUtils] [marker:invoker_docker.rm_start:23342580]
[2017-03-08T09:33:36.407Z] [INFO] [#sid_101] [SimpleExec] Running command: /usr/bin/docker rm -f wsk0_177256_limeusibmcomactioninvokenodejs1488965337811001_20170308T092858748Z
```

## Additional information you deem important:

I suspect the culprit to be in the `killStragglers` function which is called during shutdown.
It directly calls `ContainerUtils.unpauseContainer` which implies `docker unpause`. It should however call `Container.unpause` on the container instance which knows whether to unpause using docker or runc.

--- a/ansible/environments/local/group_vars/all
+++ b/ansible/environments/local/group_vars/all
@@ -30,3 +30,6 @@ limits:
 apigw_auth_user: ""
 apigw_auth_pwd: ""
 apigw_host: "http://{{ groups['apigateway']|first }}:{{apigateway.port.api}}/v1"
+
+# RunC enablement
+invoker_use_runc: true

--- a/ansible/environments/mac/group_vars/all
+++ b/ansible/environments/mac/group_vars/all
@@ -30,3 +30,6 @@ limits:
 apigw_auth_user: ""
 apigw_auth_pwd: ""
 apigw_host: "http://{{ groups['apigateway']|first }}:{{apigateway.port.api}}/v1"
+
+# RunC enablement
+invoker_use_runc: true

--- a/ansible/roles/invoker/tasks/deploy.yml
+++ b/ansible/roles/invoker/tasks/deploy.yml
@@ -28,7 +28,7 @@
 
 - name: define options when deploying invoker on Ubuntu
   set_fact:
-    linuxOptions: "-v /sys/fs/cgroup:/sys/fs/cgroup -v /run/runc:/run/runc -v {{ runcLocation | default('/usr/bin/docker-runc') }}:/usr/bin/docker-runc -v /usr/lib/x86_64-linux-gnu/libapparmor.so.1:/usr/lib/x86_64-linux-gnu/libapparmor.so.1"
+    linuxOptions: "-v /usr/lib/x86_64-linux-gnu/libapparmor.so.1:/usr/lib/x86_64-linux-gnu/libapparmor.so.1"
   when: whisk_version_name != "mac"
 
 - name: start invoker using docker cli
@@ -49,6 +49,8 @@
         -e SERVICE_CHECK_HTTP=/ping
         -e SERVICE_CHECK_TIMEOUT=2s
         -e SERVICE_CHECK_INTERVAL=15s
+        -v /sys/fs/cgroup:/sys/fs/cgroup
+        -v /run/runc:/run/runc
         -v {{whisk_logs_dir}}/invoker{{play_hosts.index(inventory_hostname)}}:/logs
         -v {{dockerInfo["json"]["DockerRootDir"]}}/containers/:/containers
         -v {{ docker_sock | default('/var/run/docker.sock') }}:/var/run/docker.sock

--- a/ansible/templates/whisk.properties.j2
+++ b/ansible/templates/whisk.properties.j2
@@ -55,7 +55,7 @@ invoker.numcore={{ invoker.numcore }}
 invoker.coreshare={{ invoker.coreshare }}
 invoker.serializeDockerOp={{ invoker.serializeDockerOp }}
 invoker.serializeDockerPull={{ invoker.serializeDockerPull }}
-invoker.useRunc={{ invoker.useRunc }}
+invoker.useRunc={{ invoker_use_runc | default(invoker.useRunc) }}
 
 consulserver.docker.endpoint={{ groups["consul_servers"]|first }}:{{ docker.port }}
 edge.docker.endpoint={{ groups["edge"]|first }}:{{ docker.port }}

--- a/core/invoker/Dockerfile
+++ b/core/invoker/Dockerfile
@@ -6,8 +6,10 @@ ENV DOCKER_VERSION 1.12.0
 # Install docker client
 RUN wget --no-verbose https://get.docker.com/builds/Linux/x86_64/docker-${DOCKER_VERSION}.tgz && \
 tar --strip-components 1 -xvzf docker-${DOCKER_VERSION}.tgz -C /usr/bin docker/docker && \
+tar --strip-components 1 -xvzf docker-${DOCKER_VERSION}.tgz -C /usr/bin docker/docker-runc && \
 rm -f docker-${DOCKER_VERSION}.tgz && \
-chmod +x /usr/bin/docker
+chmod +x /usr/bin/docker && \
+chmod +x /usr/bin/docker-runc
 
 COPY build/distributions/invoker.tar ./
 RUN tar xf invoker.tar && \

--- a/core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala
+++ b/core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala
@@ -775,7 +775,11 @@ class ContainerPool(
         val candidates = allContainers.filter { _.name.name.startsWith(actionContainerPrefix) }
         logging.info(this, s"killStragglers now removing ${candidates.length} leftover containers")
         candidates foreach { c =>
-            unpauseContainer(c.name)
+            if (useRunc) {
+                RuncUtils.resume(c.id)
+            } else {
+                unpauseContainer(c.name)
+            }
             rmContainer(c.name)
         }
         logging.info(this, s"killStragglers completed")

